### DISCUSSION TITLE
Help needed with a transform hierarchy system

### DISCUSSION THREAD

**Discussion by Boyfinn** [raysan5/raylib#3694] [Q&A] [answered]
I've been trying to implement a transform hierarchy system. So far I've only managed to copy the exact transform of an object's parent and applying it to a child of the same type.
All my objects (with and without meshes) have a `Transform `member. So far I've used the `Transform `type from raylib, used in skeletal and vertex transformations internally by raylib.
I've also tried implementing my own Transform type, with a pointer to the transform's parent, according to this:
https://gabormakesgames.com/transforms.html
My first attempt was turning a Transform into a 4x4 transformation Matrix like this:
```
Matrix MatrixFromTransform(const T3D& Trs)
{
	//Matrix translation
	Matrix _tfMatrix = MatrixIdentity();

	//Matrix scale (in local space)
	Matrix _scMatrix = MatrixScale(Trs.Scale.x, Trs.Scale.y, Trs.Scale.z);
	_tfMatrix = MatrixMultiply(_tfMatrix, _scMatrix);	//applyScale

	//Matrix rotation
	Matrix _rtMatrix = QuaternionToMatrix(Trs.Rotation);	//Convert Quaternion to rotation matrix

	Matrix _trMatrix = MatrixIdentity();	//translationMatrix
	if (Trs.Position.x != 0 || Trs.Position.y != 0 || Trs.Position.z != 0)
		_trMatrix = MatrixTranslate(Trs.Position.x, Trs.Position.y, Trs.Position.z);

	//Finalize transform
	_tfMatrix = MatrixMultiply(_tfMatrix, _rtMatrix);	//applyRotation
	_tfMatrix = MatrixMultiply(_tfMatrix, _trMatrix);	//applyTransform

	//this->_model.transform = _tfMatrix;
	return _tfMatrix;
}
```
Note that T3D here is the custom transform type with a pointer to It's parent. It's implemented pretty much identically to it's counterpart over in the blog I've linked. But just for clarity, here is what it looks like:
```
struct T3D
{
	Vector3 Position = { 0,0,0 };
	Quaternion Rotation = {0,0,0,1};
	Vector3 Scale = {1,1,1};
	T3D* Parent = nullptr;
};

```

And here is my slightly modified implementation of `GetWorldMatrix()` from the blog:
```
Matrix GetWorldMatrix(const T3D& Trs) {
	Matrix localMatrix = MatrixFromTransform(Trs);
	Matrix worldMatrix = localMatrix;

	if (Trs.Parent != nullptr) {
		Matrix parentMatrix = GetWorldMatrix(*Trs.Parent);
		worldMatrix = MatrixMultiply(localMatrix , parentMatrix);
	}
	return worldMatrix;
}
```
For every Update() call in my object tree, I call this function as such:
`this->_transform = TransformFromMatrix(GetWorldMatrix(this->_transform));`
And whenever I apply this to a 3D model, i Just update the `Model ` transform by turning it into a matrix with the` MatrixFromTransform()` function mentioned earlier.

The issue I'm having, is that all the children of an object just inherit the root object's transform directly when calling the `GetWorldMatrix()` function in this manner.
And in addition, If i try to add rotation to the transform IE: anything else than Quaternion.identity, All the objects and models just explode and fly off the screen during the first frame.
I've also tried using the blog's example of a Transform-Matrix function, but I'm not sure the order given within the blog's example applies here:
```
Matrix MatrixFromTransform(const T3D& Trs) {
	// First, extract the rotation basis of the transform
	Vector3 x = Vector3RotateByQuaternion(Vector3{ 1,0,0 }, Trs.Rotation);
	Vector3 y = Vector3RotateByQuaternion(Vector3{ 0,1,0 }, Trs.Rotation);
	Vector3 z = Vector3RotateByQuaternion(Vector3{ 0,0,1 }, Trs.Rotation);

	// Next, scale the basis vectors
	x = Vector3Scale(x , Trs.Scale.x); // Vector * float
	y = Vector3Scale(y , Trs.Scale.y); // Vector * float
	z = Vector3Scale(z , Trs.Scale.z); // Vector * float

	// Extract the position of the transform
	Vector3 t = Trs.Position;

	// Create matrix
	return Matrix{
		x.x, y.x, z.x, t.x,
		x.y, y.y, z.y, t.y,
		x.z, y.z, z.z, t.z,
		0, 0, 0, 1
	};
}
```
Otherwise, the first implementation of `MatrixFromTransform()` shown earlier, works fine for objects with no children.

I'd greatly appreciate if anyone here with better understanding of 3D transformations could help me out here.
I can provide more code/snippets if anyone needs it. Just let me know.

Thank you.

**Comment by Boyfinn** [comment, marked as answer]
I've completely neglected my global transform for each object. Closing this thread, sorry for the confusion

**Comment by coderextreme** [comment]
Hi, I’ve been thinking about implementing my own transform hierarchy.  Is there any chance of getting this into Raylib?

**Reply by Boyfinn** [reply]
https://github.com/juniper-dusk/raylib-transform
Though I'd really appreciate if this feature was added to Raylib. Also, a DrawModel function that implements quaternions as their rotation, which could be useful for teaching purposes too.
